Samples/DirectMLNpuInference/main.cpp

```diff
@@ -0,0 +1,183 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "pch.h"
+
+#include <dxcore_interface.h>
+#include <dxcore.h>
+
+#include "onnxruntime_cxx_api.h"
+#include "dml_provider_factory.h"
+
+#include "TensorHelper.h"
+
+using Microsoft::WRL::ComPtr;
+
+void InitializeDirectML(ID3D12Device1** d3dDeviceOut, ID3D12CommandQueue** commandQueueOut, IDMLDevice** dmlDeviceOut) {
+    // Whether to skip adapters which support Graphics in order to target NPU for testing
+    bool forceComputeOnlyDevice = true;
+    ComPtr<IDXCoreAdapterFactory> factory;
+    HMODULE dxCoreModule = LoadLibraryW(L"DXCore.dll");
+    if (dxCoreModule)
+    {
+        auto dxcoreCreateAdapterFactory = reinterpret_cast<HRESULT(WINAPI*)(REFIID, void**)>(
+            GetProcAddress(dxCoreModule, "DXCoreCreateAdapterFactory")
+            );
+        if (dxcoreCreateAdapterFactory)
+        {
+            dxcoreCreateAdapterFactory(IID_PPV_ARGS(&factory));
+        }
+    }
+    // Create the DXCore Adapter
+    ComPtr<IDXCoreAdapter> adapter;
+    if (factory)
+    {
+        const GUID dxGUIDs[] = { DXCORE_ADAPTER_ATTRIBUTE_D3D12_CORE_COMPUTE };
+        ComPtr<IDXCoreAdapterList> adapterList;
+        THROW_IF_FAILED(factory->CreateAdapterList(ARRAYSIZE(dxGUIDs), dxGUIDs, IID_PPV_ARGS(&adapterList)));
+        for (uint32_t i = 0, adapterCount = adapterList->GetAdapterCount(); i < adapterCount; i++)
+        {
+            ComPtr<IDXCoreAdapter> nextGpuAdapter;
+            THROW_IF_FAILED(adapterList->GetAdapter(static_cast<uint32_t>(i), IID_PPV_ARGS(&nextGpuAdapter)));
+            if (!forceComputeOnlyDevice || !nextGpuAdapter->IsAttributeSupported(DXCORE_ADAPTER_ATTRIBUTE_D3D12_GRAPHICS))
+            {
+                adapter = std::move(nextGpuAdapter);
+                break;
+            }
+        }
+    }
+    // Create the D3D12 Device
+    ComPtr<ID3D12Device1> d3dDevice;
+    if (adapter)
+    {
+        HMODULE d3d12Module = LoadLibraryW(L"d3d12.dll");
```

Comment:
> HMODULE d3d12Module = [](http://example.com/codeflow?start=7&length=22)

would be nice to print which adapter is used

Reply:
I believe that only newer Windows OS SDK versions have the DXCoreAdapter property for querying the adapter name, and I dont see it in 10.0.22621.0

Reply:
Dx dispatch prints it looks like it is in  Windows 10 (Build 18936)

[https://learn.microsoft.com/en-us/windows/win32/api/dxcore_interface/ne-dxcore_interface-dxcoreadapterproperty](https://learn.microsoft.com/en-us/windows/win32/api/dxcore_interface/ne-dxcore_interface-dxcoreadapterproperty)